File: core/src/main/java/cloud/vinh/rebirthsaga/presentation/dungeon/DungeonRenderer.java
package cloud.vinh.rebirthsaga.presentation.dungeon;

import cloud.vinh.rebirthsaga.game.grid.FloorMap;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.MathUtils;

/** World-space rendering for the prototype dungeon: terrain tiles then actors,
 * through one SpriteBatch and the world camera. Also owns the move
 * presentation track: while a committed step animates, the authoritative
 * position is already the destination; the track only interpolates the sprite
 * and never changes gameplay state. Idle frames advance only this track. */
public final class DungeonRenderer {
    public static final int TILE_SIZE = 16;
    private static final float MOVE_SECONDS = 0.12f;
    private static final float STEP_FRAME_SECONDS = 0.09f;

    private final TextureRegion floorTile;
    private final TextureRegion wallTile;
    private final TextureRegion doorTile;
    private final TextureRegion exitTile;
    private final TextureRegion[] playerFrames = new TextureRegion[2];

    private FloorMap floor;
    private int playerCellX;
    private int playerCellY;
    private final int[] moveFrom = new int[2];
    private final int[] moveTo = new int[2];
    private boolean moving;
    private float moveElapsed;
    private float frameClock;

    public DungeonRenderer(TextureAtlas atlas) {
        floorTile = requireRegion(atlas, "floor");
        wallTile = requireRegion(atlas, "wall");
        doorTile = requireRegion(atlas, "door");
        exitTile = requireRegion(atlas, "exit");
        playerFrames[0] = requireRegion(atlas, "player_a");
        playerFrames[1] = requireRegion(atlas, "player_b");
    }

    private static TextureRegion requireRegion(TextureAtlas atlas, String name) {
        TextureRegion region = atlas.findRegion(name);
        if(region == null)
            throw new IllegalArgumentException("dungeon atlas is missing region '" + name + "'");
        return region;
    }

    public void setFloor(FloorMap floor) {
        this.floor = floor;
    }

    /** Places the sprite on the authoritative cell and stops any animation. */
    public void snapPlayer(int x, int y) {
        playerCellX = x;
        playerCellY = y;
        moving = false;
        moveElapsed = 0f;
    }

    /** Begins interpolating from the previous cell to the committed cell. */
    public void beginMove(int fromX, int fromY, int toX, int toY) {
        moveFrom[0] = fromX;
        moveFrom[1] = fromY;
        moveTo[0] = toX;
        moveTo[1] = toY;
        moving = true;
        moveElapsed = 0f;
    }

    public boolean isAnimating() {
        return moving;
    }

    /** Committed (authoritative) player cell; input targeting uses this, not
     * the interpolated sprite position. */
    public int playerCellX() {
        return playerCellX;
    }

    public int playerCellY() {
        return playerCellY;
    }

    /** Animated sprite centre in world pixels; the camera follows this. */
    public float spriteCenterX() {
        return (moving ? interpolate(moveFrom[0], moveTo[0]) : playerCellX + 0.5f) * TILE_SIZE;
    }

    public float spriteCenterY() {
        return (moving ? interpolate(moveFrom[1], moveTo[1]) : playerCellY + 0.5f) * TILE_SIZE;
    }

    private float interpolate(int from, int to) {
        float progress = MathUtils.clamp(moveElapsed / MOVE_SECONDS, 0f, 1f);
        return from + (to - from) * progress;
    }

    private void update(float delta) {
        if(moving) {
            moveElapsed += delta;
            frameClock += delta;
            if(moveElapsed >= MOVE_SECONDS) {
                playerCellX = moveTo[0];
                playerCellY = moveTo[1];
                moving = false;
                moveElapsed = 0f;
            }
        }
    }

    /** Draws one frame: terrain pass, then the actor pass (stable depth rule:
     * layer first, then cell y, then stable id — one actor today). */
    public void render(SpriteBatch batch, OrthographicCamera camera, float delta) {
        update(delta);
        batch.setProjectionMatrix(camera.combined);
        batch.begin();
        if(floor != null) {
            for(int x = 0; x < floor.width(); x++) {
                for(int y = 0; y < floor.height(); y++) {
                    int tile = floor.tileAt(x, y);
                    batch.draw(floorTile, x * TILE_SIZE, y * TILE_SIZE);
                    if(tile == FloorMap.WALL)
                        batch.draw(wallTile, x * TILE_SIZE, y * TILE_SIZE);
                    else if(tile == FloorMap.DOOR)
                        batch.draw(doorTile, x * TILE_SIZE, y * TILE_SIZE);
                    else if(tile == FloorMap.EXIT)
                        batch.draw(exitTile, x * TILE_SIZE, y * TILE_SIZE);
                }
            }
        }
        TextureRegion frame = playerFrames[0];
        if(moving)
            frame = ((int)(frameClock / STEP_FRAME_SECONDS) % 2) == 0 ? playerFrames[0] : playerFrames[1];
        float drawX = spriteCenterX() - TILE_SIZE / 2f;
        float drawY = spriteCenterY() - TILE_SIZE / 2f;
        batch.draw(frame, drawX, drawY);
        batch.end();
    }
}

File: core/src/main/kotlin/cloud/vinh/rebirthdungeon/presentation/dungeon/DungeonRenderer.kt
package cloud.vinh.rebirthdungeon.presentation.dungeon

import cloud.vinh.rebirthdungeon.data.content.ContentBundle
import cloud.vinh.rebirthdungeon.game.events.ActorMoved
import cloud.vinh.rebirthdungeon.game.grid.FloorMap
import cloud.vinh.rebirthdungeon.game.identity.ContentId
import cloud.vinh.rebirthdungeon.game.projection.DungeonObservation
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import ktx.graphics.use

/** Receives observed values only. Unseen actor positions and actual hidden terrain never enter here. */
class DungeonRenderer(atlas: TextureAtlas, content: ContentBundle,
    val combatTracks: cloud.vinh.rebirthdungeon.presentation.animation.CombatTracks = cloud.vinh.rebirthdungeon.presentation.animation.CombatTracks(),
    private val font: com.badlogic.gdx.graphics.g2d.BitmapFont? = null) {
    private val regions = content.visuals.associate { binding -> binding.id to binding.frames.map { checkNotNull(atlas.findRegion(it)) } }
    private var observation: DungeonObservation? = null
    var playerCellX = 0
        private set
    var playerCellY = 0
        private set
    private var elapsed = 1f
    private var fromX = 0
    private var fromY = 0
    private var toX = 0
    private var toY = 0
    private var lastEvent = 0L
    private var runId: String? = null
    val isAnimating get() = elapsed < MOVE_SECONDS || combatTracks.playing
    val cameraFeedbackX get() = if (combatTracks.tracks.any { it.kind == "damage" && it.actor == observation?.player })
        kotlin.math.sin(combatTracks.progress * kotlin.math.PI.toFloat() * 4) * (1 - combatTracks.progress) * 2f else 0f
    val spriteCenterX get() = (fromX + (toX - fromX) * progress() + 0.5f) * TILE_SIZE
    val spriteCenterY get() = (fromY + (toY - fromY) * progress() + 0.5f) * TILE_SIZE
    private fun progress() = (elapsed / MOVE_SECONDS).coerceIn(0f, 1f)
    fun snapPlayer(x: Int, y: Int) {
        playerCellX = x; playerCellY = y; fromX = x; toX = x; fromY = y; toY = y; elapsed = MOVE_SECONDS
    }
    fun accept(value: DungeonObservation, animate: Boolean = true) {
        if (runId != value.runId) { runId = value.runId; lastEvent = 0 }
        combatTracks.accept(value, animate)
        observation = value
        val movement = value.events.filter { it.sequence > lastEvent }.map { it.event }.filterIsInstance<ActorMoved>().lastOrNull { it.actor == value.player }
        lastEvent = maxOf(lastEvent, value.events.maxOfOrNull { it.sequence } ?: 0)
        if (animate && movement != null) {
            fromX = movement.from.x; fromY = movement.from.y; toX = movement.to.x; toY = movement.to.y; elapsed = 0f
        } else snapPlayer(value.playerCell.x, value.playerCell.y)
    }
    fun skip() { combatTracks.skip(); elapsed = MOVE_SECONDS; observation?.let { snapPlayer(it.playerCell.x, it.playerCell.y) } }
    fun render(batch: SpriteBatch, camera: OrthographicCamera, delta: Float) {
        combatTracks.advance(delta)
        elapsed = minOf(MOVE_SECONDS, elapsed + delta)
        if (!isAnimating) { playerCellX = toX; playerCellY = toY }
        val view = observation ?: return
        batch.use(camera) { b ->
            for (y in 0 until view.height) for (x in 0 until view.width) {
                val tile = view.tileAt(x, y)
                if (tile < 0) continue
                val brightness = if (view.visibleAt(x, y)) 1f else 0.28f
                b.setColor(brightness, brightness, brightness, 1f)
                val id = when (tile) {
                    FloorMap.WALL -> "tile.wall"
                    FloorMap.DOOR, FloorMap.LOCKED_DOOR -> "tile.door"
                    FloorMap.EXIT -> "tile.exit"
                    else -> "tile.floor"
                }
                b.draw(regions.getValue(ContentId(id)).first(), (x * TILE_SIZE).toFloat(), (y * TILE_SIZE).toFloat())
            }
            for (actor in view.actors.sortedWith(compareBy({ it.cell.y }, { it.id.value }))) {
                val frames = regions.getValue(actor.definition)
                if (actor.player) b.setColor(1f, 1f, 1f, 1f) else b.setColor(1f, 0.45f, 0.45f, 1f)
                val movement = combatTracks.tracks.lastOrNull { it.actor == actor.id && it.kind == "move" }
                if (combatTracks.tracks.any { it.actor == actor.id && it.kind == "attack" }) b.setColor(1f, 0.85f, 0.4f, 1f)
                val x = if (actor.player) spriteCenterX - TILE_SIZE / 2f else ((movement?.from?.x?.let { it + (actor.cell.x - it) * combatTracks.progress } ?: actor.cell.x.toFloat()) * TILE_SIZE)
                val y = if (actor.player) spriteCenterY - TILE_SIZE / 2f else ((movement?.from?.y?.let { it + (actor.cell.y - it) * combatTracks.progress } ?: actor.cell.y.toFloat()) * TILE_SIZE)
                b.draw(frames[if (isAnimating) (elapsed / 0.06f).toInt() % frames.size else 0], x, y)
            }
            b.setColor(1f, 1f, 1f, 1f)
            combatTracks.tracks.filter { it.kind != "move" }.forEach { track ->
                val text = when (track.kind) { "damage" -> "-${track.amount}"; "death" -> "X"; else -> "!" }
                font?.draw(b, text, track.cell.x * TILE_SIZE.toFloat(), (track.cell.y + 1) * TILE_SIZE + combatTracks.progress * 8f)
            }
        }
    }
    companion object { const val TILE_SIZE = 16; private const val MOVE_SECONDS = 0.12f }
}
